src/includes/control-datatypes.h

```diff
@@ -0,0 +1,20 @@
+#ifndef DINIT_CONTROL_DATATYPES_H
+#define DINIT_CONTROL_DATATYPES_H
+
+#include <cstdint>
+
+namespace dinit_cptypes {
+    // A mapping between service records and their associated numerical identifier
+    // used in communction
+    using handle_t = uint32_t;
+    using trigger_val_t = uint8_t;
+    using cp_cmd_t = uint8_t;
+    using cp_rply_t = uint8_t;
+    using cp_info_t = uint8_t;
+    using srvname_len_t = uint16_t;
+    using envvar_len_t = uint16_t;
+    using sig_num_t = int32_t;
```

Comment:
This one (`sig_num_t`) isn't correct, a signal number is an `int` and not guaranteed to be 32-bit.

Reply:
Yes, I'm aware that it is not guaranteed to be 32-bit.
The question here would be do we want the protocol to be stable (even across architectures/systems) so it could be used over system boundaries (e.g. over network and/or forwarded to container/vm) ? In that case I would propose to leave it be fixed to 32-bit for the protocol and cast to int in the application. When there is a system, where int is not 32-bit then we may need to handle the conversion in dinitctl differently with an additional check, warning or error.

Reply:
I've never had the goal of having the protocol be usable over system boundaries with heterogeneous systems. That requires more than just consistently sized values; you need to ensure consistent byte order, as well, and no struct declaration will do that for you.

I know that for the most part `int32_t` will work just fine - 99% of the time it will be the case that `sizeof(int) == sizeof(int32_t)` - and the remainder of the time signal numbers will probably not be bigger than 2^31 anyway. However, although it's a bit arbitrary, "works on any POSIX system" _is_ a design goal and POSIX just says that signal numbers are `int`.

Reply:
So we should be in the clear here, as dinitctl handles the out_of_range exception if the specified signal number does not fit into a sig_num_t/int32_t. I've also added a cast from int32_t to int in control.cc, to make it more explicit. So when there is a system with different sizes it should not break, we just don't support signal numberrs >32bit atm.

Reply:
I may have been unclear. Sorry, but no - the signal number type needs to be `int`. That's what POSIX declares it as.

> as dinitctl handles the out_of_range exception if the specified signal number does not fit into a sig_num_t/int32_t

dinitctl uses `std::stoi` to convert the signal number and that returns `int`. It only throws `out_of_range` if the number doesn't fit into an `int` - it doesn't know anything about `sig_num_t`/`int32_t`. So if `int` is larger than 32-bits there's no checking done in dinitctl that the provided signal number will fit in `int32_t`. Just assigning the result to `int32_t` will silently truncate it. (Or am I missing something?)

> we just don't support signal numberrs >32bit atm

That would be a regression introduced with this change; currently, we do in fact support signal numbers >32 bit.

Everything else in the PR looks good!

Reply:
Ah.. now I get it. I've reverted sig_num_t to be int. And I've checked again, that at least in the signal protocol code sizeof is used consequently, so it doesn't break the protocol on systems where sizeof(int) != 4.